### PR TITLE
[WIP] Config Continous Integration (CCI)

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/configs/llama3/8B_qlora_single_device.yaml [hunk 1]
```diff
@@ -83,7 +83,7 @@ dtype: bf16
 
 # Activations Memory
 enable_activation_checkpointing: True
-enable_activation_offloading: True
+enable_activation_offloading: False
```
Comment (SalmanMohammadi) [inline recipes/configs/llama3/8B_qlora_single_device.yaml:86]: This is the only recipe where it was enabled.

File: tests/test_config.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        assert cfg.device == "cuda"
+        if cfg.get("enable_activation_offloading", False):
+            cfg.pop("enable_activation_offloading")
+
+        cfg.device = "meta"
+        cfg.output_dir = str(tmpdir)
+        self.validate_checkpointer(cfg.checkpointer)
+        self.validate_tokenizer(cfg.tokenizer)
+        if "fused" in cfg.optimizer:
+            cfg.optimizer.pop("fused")
```
Comment (SalmanMohammadi) [inline tests/test_config.py:109]: Recipe will error out with device meta and fused, so we validate here instead.

File: tests/test_config.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+def load_module_from_path(module_name, path):
+    spec = importlib.util.spec_from_file_location(module_name, path)
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module
+
+
+class TestRecipeConfigs:
+    def validate_tokenizer(self, tokenizer_cfg):
+        with pytest.raises(OSError, match="No such file or directory"):
```
Comment (SalmanMohammadi) [inline tests/test_config.py:51]: Actually maybe this test is stupid because it doesn't validate anything after a given OSError being raised. I should instead mock the load function and just make sure it instantiates OK.

File: tests/test_config.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    def validate_checkpointer(self, checkpointer_cfg):
+        checkpointer_class = torchtune.config._utils._get_component_from_path(
+            checkpointer_cfg["_component_"]
+        )
+        with patch(
+            "torchtune.training.checkpointing._checkpointer.get_path",
+            return_value="boo",
+        ):
+            if checkpointer_class.__name__ == "FullModelHFCheckpointer":
+                with pytest.raises(OSError, match="No such file or directory"):
```
Comment (SalmanMohammadi) [inline tests/test_config.py:63]: same as above

File: tests/test_config.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        cfg.tokenizer = OmegaConf.create(
+            {
+                "_component_": "torchtune.models.llama3.llama3_tokenizer",
+                "path": TOKENIZER_PATHS["llama3"],
+            }
+        )
+
+        with torch.device("meta"):
+            if "lora" in cfg.model._component_:
+                with patch.object(
```
Comment (SalmanMohammadi) [inline tests/test_config.py:119]: We need to do this because register_state_dict_hook attempts to copy parameters, and you can't do anything state-dependent on meta tensors.

File: tests/test_config.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                    return_value=None,
+                ):
+                    model = config.instantiate(cfg.model)
+                    state_dict = {
+                        k: v for k, v in model.state_dict().items() if "lora" not in k
+                    }
+            else:
+                state_dict = config.instantiate(cfg.model).state_dict()
+
+        load_dataset.return_value = [0]
```
Comment (felipemello1) [inline tests/test_config.py:131]: dont know what this is doing

**Review by felipemello1 — COMMENTED** [review]
thanks for this @SalmanMohammadi ! Let me see if i am getting this right: You are basically loading the config and instantiating the recipe. 

I understand having to validade the checkpointer independently, since you patch load_checkpoint, but i am not sure why you have to do it with the tokenizer, for example, if recipe.setup() will call it. Am I misunderstanding it?

Overall, i think that the test is very useful, but its not clear to me if its expensive. It was a bit hard to understand it. I think that with more structure and comments, it should be easier to review.

Please, dont spend a long time making changes. Can you do a 10~20min restructuring/commenting, and then i can come back to this?

File: tests/test_config.py (hunk 6 above)
Comment (SalmanMohammadi) [reply tests/test_config.py:131]: This mocks the HF `load_dataset` function which we don't want to make an expensive call to. I thought a bit about using the HF API to make sure the dataset exists but I felt that was a bit overkill.

**Comment by SalmanMohammadi** [comment]
RE computational cost:
```bash
============================= slowest 20 durations =============================
487.49s call     tests/recipes/test_lora_finetune_single_device.py::TestLoRAFinetuneSingleDeviceRecipe::test_loss_qlora[True-bf16]
377.84s call     tests/recipes/test_lora_finetune_single_device.py::TestLoRAFinetuneSingleDeviceRecipe::test_loss_qlora[False-bf16]
95.78s call     tests/recipes/test_lora_finetune_single_device.py::TestLoRAFinetuneSingleDeviceRecipe::test_loss_qlora[True-fp32]
68.11s call     tests/recipes/test_full_finetune_single_device.py::TestFullFinetuneSingleDeviceRecipe::test_loss[llama2/7B_full_low_memory-llama2-meta-True]
67.96s call     tests/recipes/test_knowledge_distillation_single_device.py::TestKDSingleDeviceRecipe::test_loss[qwen2/knowledge_distillation_single_device-llama3-tune-True]
61.68s call     tests/recipes/test_lora_finetune_single_device.py::TestLoRAFinetuneSingleDeviceRecipe::test_loss[llama2/7B_lora_single_device-llama2-meta-True]
61.40s call     tests/recipes/test_eleuther_eval.py::TestEleutherEval::test_torchtune_checkpoint_eval_results[truthfulqa_gen-0.1-1]
51.10s call     tests/recipes/test_eleuther_eval.py::TestEleutherEval::test_torchtune_checkpoint_eval_results[truthfulqa_gen-0.1-4]
45.85s call     tests/recipes/test_full_finetune_single_device.py::TestFullFinetuneSingleDeviceRecipe::test_loss[llama3/8B_full_single_device-llama3-tune-True]
29.40s call     tests/recipes/test_lora_finetune_single_device.py::TestLoRAFinetuneSingleDeviceRecipe::test_loss[llama3/8B_lora_single_device-llama3-tune-True]
21.03s call     tests/recipes/test_lora_dpo_single_device.py::TestLoRADPOSingleDeviceRecipe::test_training_state_on_resume[False]
20.77s call     tests/recipes/test_lora_dpo_single_device.py::TestLoRADPOSingleDeviceRecipe::test_training_state_on_resume[True]
18.60s call     tests/recipes/test_lora_finetune_distributed.py::TestLoRAFinetuneDistributedRecipe::test_training_state_on_resume[llama3/8B_lora-llama3-tune-False]
18.12s call     tests/recipes/test_knowledge_distillation_single_device.py::TestKDSingleDeviceRecipe::test_training_state_on_resume
17.10s call     tests/recipes/test_lora_finetune_distributed.py::TestLoRAFinetuneDistributedRecipe::test_training_state_on_resume[llama2/7B_lora-llama2-hf-True]
17.10s call     tests/recipes/test_lora_finetune_distributed.py::TestLoRAFinetuneDistributedRecipe::test_training_state_on_resume[llama2/7B_lora-llama2-hf-False]
16.38s call     tests/recipes/test_ppo_full_finetune_single_device.py::TestPPOFullFinetuneSingleDeviceRecipe::test_training_state_on_resume_with_optimizer_in_bwd
16.16s call     tests/recipes/test_ppo_full_finetune_single_device.py::TestPPOFullFinetuneSingleDeviceRecipe::test_training_state_on_resume
15.96s call     tests/torchtune/training/test_distributed.py::TestFullyShardState::test_qlora_state_dict
15.81s call     tests/recipes/test_eleuther_eval.py::TestEleutherEval::test_torchtune_checkpoint_eval_results[truthfulqa_mc2-0.4-4]
```
So it doesn't look like any of these tests make it to the top \#20 longest runs.

**Comment by SalmanMohammadi** [comment]
Thanks so much for looking at this @felipemello1. Sorry for the very WIP state, I've polished up a bit now.

If this approach generally looks good I'd appreciate any advice on testing distributed recipes. For DPO/KD/PPO single devices recipes I think I might need a seperate tests but I can leave that as a follow up.

**Comment by felipemello1** [comment]
i will take a closer look soon. Thanks for the changes! However, I found 3 configs that were broken. I think we will need to beef this PR up a bit, or maybe having a second PR addressing other config issues. I will link the config fix here soon
